[Merged by Bors] - feat(Tactic/Linarith): Simplex Algorithm oracle

vasnesterov opened the pull request:
- Reduce the `linarith` certificate search problem to some Linear Programming problem and implement the Simplex Algorithm to solve it. 
- Set the default oracle for `linarith` to this.
- Remove unnecessary hypotheses in `Mathlib/Analysis/Calculus/BumpFunction/FiniteDimension.lean` and `Mathlib/Analysis/Distribution/SchwartzSpace.lean` which were needed with the Fourier-Motzkin oracle.
- Adjust the definition of `CerficateOracle` to enable dot notation to choose between oracles.

This addresses #2717 and #8875 (except when the user overrides the oracle)

Also, this oracle is far more efficient:
The example below takes lots of time with the Fourier-Motzkin oracle: I waited 5 minutes and still didn't get it. But with the just implemented `Linarith.SimplexAlgo.produceCertificate` oracle the proof succeeds in less than a second.
```lean
import Mathlib.Tactic.Linarith

example (x0 x1 x2 x3 x4 : ℚ) :
  3 * x0 - 15 * x1 - 30 * x2 + 20 * x3 + 12 * x4 ≤ 0 →
  35 * x0 - 30 * x1 + 12 * x2 - 15 * x3 + 18 * x4 ≤ 0 →
  5 * x0 + 20 * x1 + 24 * x2 + 20 * x3 + 9 * x4 ≤ 0 →
  -2 * x0 - 30 * x1 + 30 * x2 - 10 * x3 - 12 * x4 ≤ 0 →
  -4 * x0 - 25 * x1 + 6 * x2 - 20 * x3 ≤ 0 →
  25 * x1 + 30 * x2 - 25 * x3 + 12 * x4 ≤ 0 →
  10 * x1 - 18 * x2 - 30 * x3 + 18 * x4 ≤ 0 →
  4 * x0 + 10 * x1 - 18 * x2 - 15 * x3 + 15 * x4 ≤ 0 →
  -4 * x0 + 15 * x1 - 30 * x2 + 15 * x3 + 6 * x4 ≤ 0 →
  -2 * x0 - 5 * x1 + 18 * x2 - 25 * x3 - 161 * x4 ≤ 0 →
  -6 * x0 + 30 * x1 + 6 * x2 - 15 * x3 ≤ 0 →
  3 * x0 + 10 * x1 - 30 * x2 + 25 * x3 + 12 * x4 ≤ 0 →
  2 * x0 + 10 * x1 - 24 * x2 - 15 * x3 + 3 * x4 ≤ 0 →
  82 * x1 + 36 * x2 + 20 * x3 + 9 * x4 ≤ 0 →
  2 * x0 - 30 * x1 - 30 * x2 - 15 * x3 + 6 * x4 ≤ 0 →
  4 * x0 - 15 * x1 + 25 * x2 < 0 →
  -4 * x0 - 10 * x1 + 30 * x2 - 15 * x3 ≤ 0 →
  2 * x0 + 6 * x2 + 133 * x3 + 12 * x4 ≤ 0 →
  3 * x0 + 15 * x1 - 6 * x2 - 15 * x3 - 15 * x4 ≤ 0 →
  10 * x1 + 6 * x2 - 25 * x3 + 3 * x4 ≤ 0 →
  -2 * x0 + 5 * x1 + 12 * x2 - 20 * x3 + 12 * x4 ≤ 0 →
  -5 * x0 - 25 * x1 + 30 * x3 - 12 * x4 ≤ 0 →
  -6 * x0 - 30 * x1 - 36 * x2 + 20 * x3 + 12 * x4 ≤ 0 →
  5 * x0 - 5 * x1 + 6 * x2 - 25 * x3 ≤ 0 →
  -3 * x0 - 20 * x1 - 30 * x2 + 5 * x3 + 3 * x4 ≤ 0 →
  False := by intros; linarith (config := {oracle := Linarith.FourierMotzkin.produceCertificate})
```

I am planning to prove the "completeness" of the oracle in the next PRs, but so far I have run a stress test on randomly generated examples of various sizes, and it seems that everything is OK.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser commented:
!bench

eric-wieser commented:
At first glance this looks great! Could you add the examples from the two issues this closes as test cases in the `test` directory? Simply creating them is enough for them to be picked up as tests.

It's unfortunate that over mathlib as a whole this has almost no performance impact (besides the outlier `Mathlib.Analysis.Normed.Group.AddCircle`), but perhaps it only makes a difference on very large states; certainly this is not an argument *against* this PR :)

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/Gauss.lean:
```diff
@@ -0,0 +1,135 @@
+def subtractRow {n m : Nat} (i j : Nat) (coef : Rat) : GaussM n m Unit := do
+  let new_row := (← get).mat[j]!.zip (← get).mat[i]! |>.map fun ⟨x, y⟩ => x - coef * y
+  let subtractedMat : Matrix n m := ⟨(← get).mat.data.set! j new_row⟩
+  set {← get with mat := subtractedMat}
```
nit: I think using `modify` here is slightly more idiomatic; perhaps as `modify ({ \. with at := subtractedMat })

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/SimplexAlgo.lean:
```diff
@@ -0,0 +1,128 @@
+      return
+    let ⟨bvarIdx, fvarIdx⟩ ← try
+      choosePivots
+    catch _ => -- unfeasible
```
I think you can write
```suggestion
    catch | .Unfeasible =>
```
here, on the premise that comments can be wrong in ways code can't.

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/SimplexAlgo.lean:
```diff
@@ -0,0 +1,128 @@
+
+/-- An exception in the `SimplexAlgoM` monad. -/
+inductive SimplexAlgoException
+| Unfeasible : SimplexAlgoException
```
The naming convention is that only types get `UpperCamelCase`, their inhabitants are `lowerCamelCase`.
```suggestion
| unfeasible : SimplexAlgoException
```

eric-wieser replied:
(though perhaps this should be `infeasible`, given [these statistics](https://books.google.com/ngrams/graph?content=infeasible%2C+unfeasible&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Cinfeasible%3B%2Cc0%3B.t1%3B%2Cunfeasible%3B%2Cc0))

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/Datatypes.lean:
```diff
@@ -0,0 +1,48 @@
+variable `bound[i]` as a linear combination of variables from `free`.
+-/
+structure Table where
+  /-- Array containig the basic variables' indexes -/
```
```suggestion
  /-- Array containing the basic variables' indexes -/
```

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/Datatypes.lean:
```diff
@@ -0,0 +1,48 @@
+dimensions of `data`), but the plan is to add some `Prop`-data and make the structure strict and
+safe.
+
+Note: we avoid using `Mathlib.Data.Matrix` because it is far more efficient to store matrix as its
```
```suggestion
Note: we avoid using the `Matrix` in the root namespace because it is far more efficient to store matrix as its
```
which will create a link in the docs to the definition.

vasnesterov commented:
@eric-wieser Thank you for your feedback! I have addressed your comments and slightly revised the code, making it shorter and (I hope) easier to read.

semorrison commented on Mathlib/Tactic/Linarith/SimplexAlgo/PositiveVector.lean:
```diff
@@ -0,0 +1,80 @@
+
+`linarith` certificate search can easily be reduced to this LP problem: given the matrix `A` and the
+list `strictIndexes`, find the non-negative vector `v` such that some of its coordinates from
+the `strictIndexes` are posivie and `A v = 0`.
```
```suggestion
the `strictIndexes` are positive and `A v = 0`.
```

semorrison commented on Mathlib/Tactic/Linarith/Verification.lean:
```diff
@@ -202,7 +202,7 @@ def proveFalseByLinarith (cfg : LinarithConfig) : MVarId → List Expr → MetaM
       trace[linarith.detail] "... finished `linearFormsAndMaxVar`."
       trace[linarith.detail] "{comps}"
-      let oracle := cfg.oracle.getD FourierMotzkin.produceCertificate
+      let oracle := cfg.oracle.getD SimplexAlgo.produceCertificate
```
Could we have a comment here reminding the reader of the existence of FourierMotzkin?

(For maintainability, I'd like to be sure we can always "fall back" quickly and easily.)

eric-wieser replied:
I put a comment in the docstring

semorrison commented on Mathlib/Tactic.lean:
```diff
@@ -99,6 +99,10 @@ import Mathlib.Tactic.Linarith.Frontend
 import Mathlib.Tactic.Linarith.Lemmas
 import Mathlib.Tactic.Linarith.Parsing
 import Mathlib.Tactic.Linarith.Preprocessing
+import Mathlib.Tactic.Linarith.SimplexAlgo.Datatypes
```
Could you rename `SimplexAlgo` to either `Simplex` or `SimplexAlgorithm`?

eric-wieser commented on Mathlib/Tactic/Linarith/SimplexAlgo/Gauss.lean:
```diff
@@ -0,0 +1,125 @@
+/-- The monad for the Gaussian Elimination algorithm. -/
+abbrev GaussM (n m : Nat) := StateM <| GaussState n m
+
+/-- Gets current row. -/
+def curRow {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentRow
+
+/-- Gets current column. -/
+def curCol {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentColumn
+
+/-- Increments current row. -/
+def incRow {n m : Nat} : GaussM n m Unit :=
+  modify fun s => {s with currentRow := s.currentRow + 1}
+
+/-- Increments current column. -/
+def incCol {n m : Nat} : GaussM n m Unit :=
+  modify fun s => {s with currentColumn := s.currentColumn + 1}
+
+/-- Finds the first row starting from the current column with nonzero element in current column. -/
+def findNonzeroRow {n m : Nat} : GaussM n m <| Option Nat := do
+  for i in [← curRow:n] do
+    if (← get).mat[i]![← curCol]! != 0 then
+      return i
+  return .none
+
```
I'm not convinced this `GaussM` monad is pulling its weight here; I'd be inclined to drop all these row operations, and:

* Make `findNonzeroRow` take `curRow` and `curCol`
* Use `let mut row` and `let mut col` in `getTableImp`

Happy to let another reviewer override me.

semorrison commented on Mathlib/Tactic/Linarith/Verification.lean:
```diff
@@ -202,7 +202,7 @@ def proveFalseByLinarith (cfg : LinarithConfig) : MVarId → List Expr → MetaM
       trace[linarith.detail] "... finished `linearFormsAndMaxVar`."
       trace[linarith.detail] "{comps}"
-      let oracle := cfg.oracle.getD FourierMotzkin.produceCertificate
+      let oracle := cfg.oracle.getD SimplexAlgorithm.produceCertificate
```
```suggestion
      -- If needed, we can switch back to the Fourier-Motzkin generator using
      -- let oracle := cfg.oracle.getD FourierMotzkin.produceCertificate
      let oracle := cfg.oracle.getD SimplexAlgorithm.produceCertificate
```

eric-wieser replied:
I don't think we need this; we can already switch back with `linarith (config := { oracle := some .fourierMotzkin})` (after the changes I just pushed to fix dot notation.

semorrison commented:
Any further review comments? I think we can get this in soon.

eric-wieser commented on test/linarith.lean:
```diff
@@ -437,6 +438,11 @@ example (p q r s t u v w : ℕ) (h1 : p + u = q + t) (h2 : r + w = s + v) :
   p * r + q * s + (t * w + u * v) = p * s + q * r + (t * v + u * w) :=
 by nlinarith
 
+-- note: much faster than simplex
+example (p q r s t u v w : ℕ) (h1 : p + u = q + t) (h2 : r + w = s + v) :
+  p * r + q * s + (t * w + u * v) = p * s + q * r + (t * v + u * w) :=
+by nlinarith (config := { oracle := some .fourierMotzkin })
```
The new algorithm is about 5x slower here than the old one was.

vasnesterov replied:
It's likely due to the large size of the matrix (about 100 rows and columns), which is actually sparse (only 6% of the entries are non-zero). This is typical of `nlinarith`, and, in the next PR, I'll make it possible to use sparse matrices in the algorithm.

eric-wieser replied:
Can you edit the comment in this file to explain that simplex is likely slower here due to the use of dense matrices?

vasnesterov replied:
Done :)

eric-wieser commented:
I pushed some changes to avoid putting the `SimplexAlgorithm` stuff in a file about `FourierMotzkin`, and to make it slightly more ergonomic to switch between the two.

I added a test that shows that actually in some cases the new algorithm is worse; though I think these are relatively rare.

eric-wieser commented:
bors d=@semorrison

eric-wieser commented:
!bench

eric-wieser commented:
!bench

vasnesterov commented:
!bench

semorrison commented:
bors merge
